### DISCUSSION TITLE
libcef.dll move to c drive or gac

### DISCUSSION THREAD

**Discussion by dbdariusz** [cefsharp/CefSharp#3565] [Q&A]
Hello. I am thinking if there is a way to install libcef.dll to gac or move that dll somewhere else e.g. to c:\ drive. My goal is to reduce deployment size. That dll weight more than 100mb what has huge impact to release.
I tried use CefLibraryHandle and provide path but it does not work. 
Got error: Cef.Initialize() failed.Check

I also tried cmd: gacutil -i libcef.dll
Failure adding assembly to the cache:   The module was expected to contain an assembly manifest.

**Comment by amaitland** [comment]
Moving just `libcef.dll` is impractical, it relies on all the `bin/pak` files having a relative location. 

> I also tried cmd: gacutil -i libcef.dll
> Failure adding assembly to the cache: The module was expected to contain an assembly manifest.

`libcef.dll` is an unmanaged dll and cannot be loaded into the `GAC`.

You can specify `<CefSharpTargetDir/>` in your project file, the `CEF` resources (dlls/bin/pak) files will then be copied into a subfolder. You can then move this folder basically anywhere you like. You then just have to load `CefSharp.Core.Runtime.dll` before you attempt to access any of the `CefSharp` dlls. You can manually load `CefSharp.Core.Runtime.dll` or you can use [AppDomain.AssemblyResolve](https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.assemblyresolve?view=net-5.0). AppDomain.AssemblyResolve is often used when targeting `AnyCPU` and there's example at https://github.com/cefsharp/CefSharp.MinimalExample/blob/cefsharp/90/CefSharp.MinimalExample.WinForms/Program.cs#L20

---

Manually edit your project file e.g:
1. Edit `CefSharp.MinimalExample.WinForms.csproj` and add following code to the first `<PropertyGroup/>` or add your own `<PropertyGroup/>`  before the `.targets` files in your project.
```xml
    <CefSharpTargetDir>cefsharp</CefSharpTargetDir>
```

**Reply by dbdariusz** [reply]
Interesting. It means I can move cefsharp to e.g. to c:\ and next adjust resolver as in an example but for all missing DLLs or I just need to do it for CefSharp.Core.Runtime.dll?

Do I need to use CefLibraryHandle and adjust CefSettings for paths?

**Reply by amaitland** [reply]
> adjust resolver as in an example but for all missing DLLs or I just need to do it for CefSharp.Core.Runtime.dll?

Only `CefSharp.Core.Runtime.dll`.

> Do I need to use CefLibraryHandle and adjust CefSettings for paths?

No

**Reply by dbdariusz** [reply]
I added resolver and moved cefsharp to c:\ and got an exception during loading runtime.dll. What I missed?

```
Could not load file or assembly 'CefSharp.Core.Runtime.dll' or one of its dependencies. The specified module could not be found.
===STACK TRACE===
   at CefSharp.CefSettingsBase..ctor()
   at CefSharp.WinForms.CefSettings..ctor()


AddHandler AppDomain.CurrentDomain.AssemblyResolve, AddressOf Resolver

Private Function Resolver(ByVal sender As Object, ByVal args As ResolveEventArgs) As System.Reflection.Assembly
        If args.Name.StartsWith("CefSharp.Core.Runtime") Then
            Dim assemblyName As String = args.Name.Split({","c}, 2)(0) & ".dll"
            Dim archSpecificPath As String = Path.Combine("c:\cefsharp\9065\", assemblyName)
            Return If(File.Exists(archSpecificPath), Reflection.Assembly.LoadFile(archSpecificPath), Nothing)
        End If
        Return Nothing
End Function
```

**Reply by amaitland** [reply]
Is your Resolver function called? If not then your likely subscribing to the event handler too late.

**Reply by dbdariusz** [reply]
Yes. Resolver is called right after method Cef.IsInitialized is executed and exception is throw when Reflection.Assembly.LoadFile try to load direct path to dll (c:\cefsharp\9065\CefSharp.Core.Runtime.dll).

**Reply by amaitland** [reply]
Is libcef.dll and all the other dll/bin/pak files located in c:\cefsharp\9065\

The exception suggests there's some files missing 

https://github.com/cefsharp/CefSharp/wiki/Output-files-description-table-%28Redistribution%29

**Reply by dbdariusz** [reply]
cefsharp folder does not contain:
CefSharp.Core.dll
CefSharp.WinForms.dll
I moved these dlls manually but still the same exception.

Before I moved these above dlls folder looks as below:
![image](https://user-images.githubusercontent.com/3390787/118882406-293d9500-b8c2-11eb-9d9d-aa22a51049ce.png)

**Reply by amaitland** [reply]
> I moved these dlls manually but still the same exception.

If you move these files you will have to update your Assembly Resolve. They are direct references to your project and not copied to the  CefSharpTargetDir folder. This is expected. 

I'd suggest getting it working as a subfolder before attempting to move the folder outside the directory tree, this is the tested scenario.

**Reply by amaitland** [reply]
Using https://github.com/cefsharp/CefSharp.MinimalExample as a base

- Add `<CefSharpTargetDir>runtime</CefSharpTargetDir>` to  first property group of `CefSharp.MinimalExample.WinForms.csproj` file
- Modify `Program.cs` (code example below)
- Manually copy the contents of the `runtime` folder to `C:\projects\temp\cefsharp.runtime`.

```c#
AppDomain.CurrentDomain.AssemblyResolve += Resolver;

private static System.Reflection.Assembly Resolver(object sender, ResolveEventArgs args)
{
	if (args.Name.StartsWith("CefSharp.Core.Runtime"))
	{
		string assemblyName = args.Name.Split(new[] { ',' }, 2)[0] + ".dll";
		string path = Path.Combine(@"C:\projects\temp\cefsharp.runtime", assemblyName);

		return File.Exists(path)
				   ? System.Reflection.Assembly.LoadFile(path)
				   : null;
	}

	return null;
}
```

Application runs as expected.

**Reply by dbdariusz** [reply]
Weird. I run example app and it works. Let you know once figure out what cause it. So far I know example is in c#, my in in vb. Thanks for the help.

**Reply by dbdariusz** [reply]
I figured out. Reason of that was a security check for ClickOnce. Cannot be enable.

![image](https://user-images.githubusercontent.com/3390787/119003084-728df300-b95b-11eb-8818-601c583274fc.png)

**Reply by amaitland** [reply]
Supporting and developing CefSharp takes a great deal of time and effort, if you are using CefSharp in a commercial application please consider funding my efforts. Even $5 a month makes a difference (Paying for twelve months in advance if you wish to make a single payment).

https://github.com/sponsors/amaitland/
